mchess-master/mchess-master/CBoard.cc

```diff
@@ -316,52 +311,76 @@ void CBoard::find_legal_moves(CMoveList &moves) const
+                        iscaptured=1;
                     }
-
-                    j = i + 11; // Diagonal capture
-                    if ((m_board[j] != IV && m_board[j] < 0))
+                    j = i + 10; // One square forward
+                    if (m_board[j] == EM &&iscaptured==0)
                     {
                         if (i > 80) // Check for promotion
                         {
                             {
-                                CMove move(piece, i, j, m_board[j], WQ);
+                                CMove move(piece, i, j, EM, WQ);
                                 moves.push_back(move);
                             }
                             {
-                                CMove move(piece, i, j, m_board[j], WR);
+                                CMove move(piece, i, j, EM, WR);
                                 moves.push_back(move);
                             }
                             {
-                                CMove move(piece, i, j, m_board[j], WB);
+                                CMove move(piece, i, j, EM, WB);
                                 moves.push_back(move);
                             }
                             {
-                                CMove move(piece, i, j, m_board[j], WN);
+                                CMove move(piece, i, j, EM, WN);
                                 moves.push_back(move);
                             }
                         }
-                        else
+                        else // regular pawn move
                         {
-                            CMove move(piece, i, j, m_board[j]);
+                            CMove move(piece, i, j, EM);
                             moves.push_back(move);
                         }
+
+                        j = i + 20; // Two squares forward
+                        if (m_board[j] == EM)
+                        {
+                            if (i < 40) // Only from second rank
+                            {
+                                CMove move(piece, i, j, EM);
+                                moves.push_back(move);
+                            }
+                        }
                     }
+
+                    
                     break;
 
                 case WN : // White knight 
                     {
+                        
                         int dirs[8] = {NNW, NNE, NWW, NEE, SSW, SSE, SWW, SEE};
-
+                        for (int k=0; k<8; ++k){
+                            int dir = dirs[k];
+                            j = i+dir;
+                            if (m_board[j] != IV) 
+                                if (m_board[j] <0){
+                                    iscaptured=1;
+                                    CMove move(piece, i, j, m_board[j]);
+                                    moves.push_back(move);
+                                }
+                        }
+                        if(iscaptured==1) break;
```

Comment:
You can't break prematurely, you need to find **ALL** the captures and can't just assume we can break once 1 single capture is found